### PR TITLE
noescape by default: deprecate @noescape and @autoclosure(escaping)

### PR DESCRIPTION
**Pull request by milseman**
#### What's in this pull request?

Support the @autoclosure @escaping syntax, deprecate @noescape and @autoclosure(escaping). Fixits and tests included.
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,14 @@ Note: This is in reverse chronological order, so newer entries are added to the
 
 Swift 3.0
 ---------
+
+* [SE-103](https://github.com/apple/swift-evolution/blob/master/proposals/0103-make-noescape-default.md)
+
+  Closure parameters are non-escaping by default, rather than explicitly being
```
Comment (slavapestov) [inline CHANGELOG.md]: Bits of code in the changelog should be surrounded by ` and `

File: lib/AST/ASTPrinter.cpp [hunk 2]
```diff
@@ -3746,7 +3746,7 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
       if (info.isNoEscape())
         Printer << "@autoclosure ";
       else
-        Printer << "@autoclosure(escaping) ";
+        Printer << "@autoclosure @escaping ";
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Wouldn't it be better to just handle @autoclosure and @escaping separately?

File: lib/AST/TypeRepr.cpp [hunk 3]
```diff
@@ -276,13 +276,14 @@ void AttributedTypeRepr::printAttrs(llvm::raw_ostream &OS) const {
...
   const TypeAttributes &Attrs = getAttrs();
 
-  switch (Attrs.has(TAK_autoclosure)*2 + Attrs.has(TAK_noescape)) {
-  case 0: break;  // Nothing specified.
-  case 1: Printer << "@noescape "; break;
-  case 2: Printer << "@autoclosure(escaping) "; break;
-  case 3: Printer << "@autoclosure "; break;
+  if (Attrs.isDeprecatedAutoclosureEscaping) {
+    // Deprecated:
+    Printer << "@autoclosure(escaping) ";
```
Comment (slavapestov) [inline lib/AST/TypeRepr.cpp]: We don't need to print TypeReprs with deprecated stuff in them, I think.

File: lib/Sema/TypeCheckType.cpp [hunk 4]
```diff
@@ -1889,15 +1889,14 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
...
-    // special autoclosure knowledge to just as "isEscaping" or similar.
-    if (isFunctionParam && !attrs.has(TAK_autoclosure)) {
-      // Closure params default to non-escaping
-      if (attrs.has(TAK_noescape)) {
-        // FIXME: diagnostic to tell user this is redundant and drop it
-      } else if (!attrs.has(TAK_escaping)) {
-        defaultNoEscape = isDefaultNoEscapeContext(DC);
-      }
+    if (isFunctionParam && !attrs.has(TAK_escaping) &&
+        !attrs.isDeprecatedAutoclosureEscaping) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckType.cpp]: You can eliminate the need to special-case  attrs.isDeprecatedAutoclosureEscaping here by also setting TAK_escaping when you set attrs.isDeprecatedAutoclosureEscaping

File: lib/Sema/TypeCheckType.cpp [hunk 5]
```diff
@@ -1889,15 +1889,14 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
...
+      defaultNoEscape = isDefaultNoEscapeContext(DC);
+    }
+
+    if (isFunctionParam && attrs.has(TAK_noescape) &&
```
Comment (slavapestov) [inline lib/Sema/TypeCheckType.cpp]: Is this FIXME still relevant now that @noescape itself is deprecated?

File: test/decl/overload.swift [hunk 6]
```diff
@@ -227,7 +227,7 @@ func throwsFuncParam(_ fn: () throws -> ()) { }
 func throwsFuncParam(_ fn: () -> ()) { }
 
 // @noescape
-func noescape(x: @noescape (Int) -> Int) { } // expected-note{{previously declared}}
+func noescape(x: (Int) -> Int) { } // expected-note{{previously declared}}
 func noescape(x: (Int) -> Int) { } // expected-error{{invalid redeclaration of 'noescape(x:)'}}
```
Comment (slavapestov) [inline test/decl/overload.swift]: The goal of this test is to ensure that you cannot overload on @escaping, so you should update it. Rename noescape() to escaping(), and make one of the versions put @escaping on x

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by tkremenek** [comment]
@swift-ci test linux

File: lib/Sema/TypeCheckType.cpp (hunk 5 above)
Comment (milseman) [inline lib/Sema/TypeCheckType.cpp]: The FIXME is handled in the parser itself when it encounters `@autoclosure`.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
Failure unrelated:
Serialization/failed-clang-module.swift
/Users/buildnode/jenkins/workspace/swift-PR-osx/buildbot_incremental/swift-macosx-x86_64/test-iphonesimulator-i386/Serialization/Output/failed-clang-module.swift.tmp/MixModA.framework/Modules/module.modulemap:3:10: error: inferred submodules require a module with an umbrella
  module * { export * }
         ^

Which is a new test file that's been problematic:

**Comment by slavapestov** [comment]
@milseman Thanks for taking the time to do the changes, the patch looks much cleaner now!

**Pull request merged**